add icon classes and replace src-matching by class-based matching

clouless opened the pull request:
### Improvement

This is a suggestion to add `class="icon-fav-active"` and `class="icon-fav-inactive"` to the `<img` and replace the JS functionality that matches for `src=star.png` by a classname matching as it is done with all other icons in jenkins.

I need this for my theme development and I think this would be an improvement

**UPDATE**

 * I compiled it and tested in Jenkins v 2.32 see preview below
 * It works as expected
 * (And I fixed the built in `Favorites.java` which already had an error I did not cause)

**UPDATE 2**

 * The icon size classess are also set with `icon-md`, `icon-lg` a.s.o
 * Event is dispatched when fav-icon changes from active to inactive and vice versa
   * `window.dispatchEvent(new CustomEvent('favorite-plugin-icon-change', { transition: "inactive-to-active" }));`
   * With this code you can subscribe to the event 

```
window.addEventListener("favorite-plugin-icon-change", function(){
    console.log('EVENT RECORDED');
});
```

### Preview

![favorite-with-class](https://cloud.githubusercontent.com/assets/12599965/20459243/78e7a120-aeba-11e6-904f-ff904eefc303.gif)

i386 commented:
@clouless thanks for contributing!

I think this change could be cleaner if you replaced the `img` tag with a `div` and had CSS setting the background of the div. That way we can remove all the code that swaps out the `img src=`

i386 commented on src/main/java/hudson/plugins/favorite/Favorites.java:
```diff
@@ -136,6 +136,10 @@ public boolean hasNext() {
+
+                    /* Java 7 compatibility fallback */
+                    @Override
+                    public void remove() { throw new UnsupportedOperationException(); }
```
Why is this change necessary?

clouless replied:
The built failed because of that. And after some googling it seems that depending on the JDK that builds the plugin this is needed. When built with JDK 8 it is not necessary.
When building with JDK 7 and lesser it is. So either fix the jenkins or keep this :)

i386 replied:
Can you please submit this change as a different PR?

i386 replied:
Great I'll approve it straight away

clouless replied:
done with #7

clouless commented:
@i386 ok nice. I will change it to div with background tomorrow. Since this will involve more testing on all major browsers. I will report back here :)

i386 commented:
@clouless great! I think you can get rid of `getSizeClassFromIconSize` stuff too and just serve the largest asset and let CSS scale it appropriately. Even delete the smaller assets.

clouless commented:
@i368 Yes only one big asset would be fine. But i would like to keep the `getSizeClassFromIconSize` thing since this gives me the correct size in compliance with other jenkins icons. These classes can then be used to scale the big asset by CSS.

In the classic Jenkins you have the "S", "M" and "L" size switcher below all the jobs and the logic is to set these classnames on icons. I think that is a good convention which only the favorite-plugin does not seem to follow. So mainly I want it to follow the convention (even if the classes would be set and have no further use other than fullfilling the convention)

clouless commented:
I will need to refactor the plugin even further. I will do:

 * Create `css.jelly` like in https://github.com/jenkinsci/sloccount-plugin/tree/master/src/main/resources/tabview
 * Create `javascript.jelly` with JS and move i18n to *Message_*.properties*
 * Remove *.js files with duplicated code

clouless commented:
**UPDATE** 

  * I did all that and fixed i18n.
  * I will now merge master to resolve conflicts.

@i386 I added a README.md and a LICENSE.md with MIT License. This should be done for all plugin repos btw. :D

i386 commented on LICENSE.md:
```diff
@@ -0,0 +1,9 @@
+# The MIT License (MIT)
```
@larrys is this the original licence?

clouless replied:
ok I will remove all license stuff and move to separate PR.

i386 commented on LICENSE.md:
```diff
@@ -0,0 +1,9 @@
+# The MIT License (MIT)
+
+Copyright (c) `2016` `CloudBees, Inc.`
```
🐛  This copyright does not belong to CloudBees

Perhaps
```
Copyright (c) 2010 Larry Shatzer, Jr.
```

i386 replied:
Lets move LICENSE changes to another PR so I can get feedback from @larrys (it's his IP)

clouless replied:
ok I will remove all license stuff and move to separate PR.

i386 commented on pom.xml:
```diff
@@ -16,6 +16,12 @@
     <name>Favorite</name>
     <url>http://wiki.jenkins-ci.org/display/JENKINS/Favorite+Plugin</url>
 
+    <licenses>
```
Please remove all the license attribution changes, including the LICENSE.md and put in another PR until I can get @larrys to tell us the license.

clouless replied:
ok I will remove all license stuff and move to separate PR.

i386 commented on src/main/webapp/favoriteToggle.js:
```diff
@@ -1,12 +1 @@
-  }
-  return false;
-}
+/* MOVED: src/main/resources/hudson/plugins/favorite/assets/javascript.jelly */
```
Just delete the file

clouless replied:
ok

i386 commented on src/main/resources/hudson/plugins/favorite/assets/css.jelly:
```diff
@@ -0,0 +1,49 @@
+<?jelly escape-by-default='true'?>
```
The favorite star is not vertically aligned

![dashboard jenkins 2016-11-22 08-16-33](https://cloud.githubusercontent.com/assets/50156/20501016/1057fe74-b08c-11e6-991e-e399a69941af.png)

clouless replied:
fixed. icons align vertical now.

i386 replied:
Verified here - thanks!

i386 commented on pom.xml:
```diff
@@ -71,8 +77,18 @@
     </dependencies>
-
+    <properties>
+        <jenkins.version>2.32</jenkins.version>
```
Please remove the version as it would make this plugin incompatible with Blue Ocean which needs 2.7.1 or above.

If there is a good reason for doing so (for example, CSS or JavaScript needs Jenkins 2) then set this to 2.7.1.

clouless replied:
I just moved this stuff inside because `mvn:hpi run` had problems otherwise. But I could start `mvn hpi:run` with parameters instead of defining it in the pom.

i386 commented on pom.xml:
```diff
@@ -71,8 +77,18 @@
-
+    <properties>
+        <jenkins.version>2.32</jenkins.version>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
```
This comes from the base POM. Please remove.

i386 commented on pom.xml:
```diff
@@ -71,8 +77,18 @@
+    <properties>
+        <jenkins.version>2.32</jenkins.version>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <hpi-plugin.version>1.120</hpi-plugin.version>
```
Also comes from the parent pom. Please remove.

i386 commented on pom.xml:
```diff
@@ -71,8 +77,18 @@
             <version>1.5.1</version>
             <optional>true</optional>
         </dependency>
+        <dependency>
```
Surprised that you needed a new JUnit. Why is this needed? We should always use the version in the parent pom

The version the parent POM lists is:
```
     <dependency>
        <groupId>junit</groupId>
        <artifactId>junit</artifactId>
        <version>4.12</version>
      </dependency>
```

clouless replied:
I had problems when running `mvn hpi:run` telling me I needed this, but I can remove it.

i386 commented on src/main/java/hudson/plugins/favorite/Favorites.java:
```diff
@@ -136,7 +136,7 @@ public boolean hasNext() {
                         return jenkins.getItemByFullName(iterator.next());
                     }
-                    
+
```
Undo this whitespace change

clouless replied:
ok

i386 commented on src/main/resources/hudson/plugins/favorite/column/FavoriteColumn/column.jelly:
```diff
@@ -3,10 +3,11 @@
     <td data="${sortData}">
       <a onclick="return toggleFavorite('${job.getFullName()}', '${id.userId}', this)" href="#">
-        <img src="${imagesURL}/${subIconSize}/${star}" border="0" id="fav_${job.getFullName()}"/>
+        <span id="fav_${job.getFullName()}" class="${starClass} ${sizeClass}"></span>
```
Nice 💯

clouless commented:
@i386 I fixed all requested things. Let's do the LICENSE stuff later when this PR is fully merged.

Sorry for the Time-delay. Being in Germany and you seem to be in Australia makes the communication very asynchronous. I hope that is ok :)

i386 commented:
I used `mvn hpi:run`

i386 commented:
This LGTM 💯

i386 commented:
Can you please squash all the commits and rebase against master?

clouless commented:
@i386 ok will do. Do you have any commandline git commands for me? I don't want to mess it up ^^

i386 commented:
I merged this in #9

i386 commented:
Released as favorite 2.0.3 - thanks!

clouless commented:
Thanks :) And great job with blue ocean btw. I use it a lot and will write a blogpost about it on wednesday :)

i386 commented:
@clouless thank you! Oooh do link me the blog post when its up 👍

i386 commented:
@clouless looks like this changed introduced a bug. Would you mind taking a look? https://issues.jenkins-ci.org/browse/JENKINS-39990

clouless commented:
@i386 sure. I will provide a new PR with a fix right away.

i386 commented:
@clouless thanks for jumping on this so quick 🎉

clouless commented:
@i386 no problem :D sorry for breaking it 😼
